### ISSUE TITLE
What sparked your interest in learning about parsers and ultimately creating Esprisma?

### ISSUE THREAD

**Issue by cem2ran** [ariya/ama#28]
Alternative title: What was the motivation behind Esprisma?

**Comment by ariya** [comment]
**Necessity is the mother of invention**.

Back in 2011, I wanted to perform a static analysis on JavaScript programs using JavaScript. Before I [announced](http://ariya.ofilabs.com/2011/12/introducing-esprima.html) Esprima, my choices were pretty limited. I could use **JSLint** but its parser was (and still is) too intertwined with its linting code. An alternative would be **UglifyJS** (the old one, not the recent version 2), however its syntax format was essentially "values in array" and it was very hard to work with. Finally there was **Reflect.js**, but at that time it was not mature enough, failing some real-world libraries I threw at it. There were some other alternatives I have explored but every one of them did not pass my criteria:
- Specification-compliant, as close as possible to the official ECMA-262
- Reasonably fast, I couldn't wait one second to parse the entire jQuery code
- Sensible syntax tree format

This was why Esprima was born. I am not a computer scientist (see #21 for the details) so I just picked a syntax tree format resembling the [SpiderMonkey Parser API](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/Parser_API) (now consolidated as [ESTree](https://github.com/estree/estree/)). Performance is important, during Esprima development I investigated and documented a few JavaScript [performance tricks](http://ariya.ofilabs.com/2014/04/supersonic-javascript.html), from [fixed object structure](http://ariya.ofilabs.com/2012/02/javascript-object-structure-speed-matters.html), [profile-guided reordering](http://ariya.ofilabs.com/2013/07/profile-guided-javascript-optimization.html) to [trie-like decision speed-up](http://ariya.ofilabs.com/2012/02/javascript-object-structure-speed-matters.html). While it is far from perfect, Esprima is equipped with [all kind of tests](http://esprima.org/test/ci.html) to ensure its compliance and correctness, from over 1,300 unit tests to various forms of regression tests. And of course, one line item in its [release checklist](https://github.com/jquery/esprima/wiki/Release-Preparation) is to verify that the performance does not regress.

In short, I built it because I needed to scratch my own itch.

Thank you @cem2ran!
